File: rust/src/bin/day10.rs
use std::{
    collections::{BTreeSet, VecDeque},
    io::BufRead,
};

use anyhow::Result;

use rust::read;

#[derive(Debug)]
struct Map {
    data: Vec<Vec<u8>>,
}

impl Map {
    fn zeros(&self) -> Vec<(usize, usize)> {
        self.data
            .iter()
            .enumerate()
            .flat_map(|(x, line)| {
                line.iter()
                    .enumerate()
                    .filter(|(_, c)| c == &&0)
                    .map(|(y, _)| (x, y))
                    .collect::<Vec<(usize, usize)>>()
            })
            .collect()
    }
    fn adj_points(&self, (x, y): (usize, usize)) -> Vec<(usize, usize)> {
        let (x, y) = (x as i32, y as i32);
        let pos = [(x + 1, y), (x, y + 1), (x - 1, y), (x, y - 1)];

        let good = |(x, y): (&i32, &i32)| {
            *x >= 0 && *x < self.data.len() as i32 && *y >= 0 && *y < self.data[0].len() as i32
        };

        pos.into_iter()
            .filter(|(x, y)| good((x, y)))
            .map(|(x, y)| (x as usize, y as usize))
            .collect()
    }

    fn count_score(&self, pos: (usize, usize)) -> usize {
        let mut visited: BTreeSet<(usize, usize)> = BTreeSet::new();
        let mut queue = VecDeque::new();

        queue.push_back(pos);
        while let Some(pos) = queue.pop_front() {
            if visited.contains(&pos) {
                continue;
            }
            visited.insert(pos);
            let value = self.data[pos.0][pos.1];
            let good_points = self
                .adj_points(pos)
                .into_iter()
                .filter(|p| !visited.contains(p))
                .filter(|(x, y)| value + 1 == self.data[*x][*y]);
            queue.extend(good_points);
        }

        visited
            .into_iter()
            .filter(|(x, y)| self.data[*x][*y] == 9)
            .count()
    }

    fn count_score_part2(&self, pos: (usize, usize)) -> usize {
        let mut visited: BTreeSet<(usize, usize)> = BTreeSet::new();
        let mut score_map = vec![vec![0; self.data[0].len()]; self.data.len()];
        let mut queue = VecDeque::new();

        score_map[pos.0][pos.1] = 1;
        queue.push_back(pos);
        while let Some(pos) = queue.pop_front() {
            if visited.contains(&pos) {
                continue;
            }
            visited.insert(pos);
            let value = self.data[pos.0][pos.1];
            let score = score_map[pos.0][pos.1];
            self.adj_points(pos)
                .into_iter()
                .filter(|p| !visited.contains(p))
                .filter(|(x, y)| value + 1 == self.data[*x][*y])
                .for_each(|(x, y)| {
                    score_map[x][y] += score;
                    queue.push_back((x, y));
                });
        }
        visited
            .into_iter()
            .filter(|(x, y)| self.data[*x][*y] == 9)
            .map(|(x, y)| score_map[x][y])
            .sum()
    }
}

fn read_map() -> Result<Map> {
    let reader = read(10, false)?;
    let data = reader
        .lines()
        .map_while(Result::ok)
        .filter(|s| !s.is_empty())
        .map(|s| s.split("").filter_map(|c| c.parse::<u8>().ok()).collect())
        .collect();
    Ok(Map { data })
}

fn part1() -> Result<()> {
    let map = read_map()?;
    let zeros = map.zeros();
    let ans: usize = zeros.into_iter().map(|p| map.count_score(p)).sum();
    println!("{}", ans);
    Ok(())
}

fn part2() -> Result<()> {
    let map = read_map()?;
    let zeros = map.zeros();
    let ans: usize = zeros.into_iter().map(|p| map.count_score_part2(p)).sum();
    println!("{}", ans);
    Ok(())
}

fn main() -> Result<()> {
    part1()?;
    part2()?;
    Ok(())
}

File: python/day10.py
import dataclasses
from collections import defaultdict
from functools import cached_property

import utils


@dataclasses.dataclass(unsafe_hash=True)
class Point:
    x: int
    y: int


@dataclasses.dataclass
class Map:
    data: list[list[int]]

    @cached_property
    def h(self) -> int:
        return len(self.data)

    @cached_property
    def w(self) -> int:
        return len(self.data[0])

    def zeros(self) -> list[Point]:
        return [
            Point(x, y)
            for x, line in enumerate(self.data)
            for y, c in enumerate(line)
            if c == 0
        ]

    def adjacents(self, point: Point) -> list[Point]:
        def good(p: Point) -> bool:
            return p.x >= 0 and p.x < self.h and p.y >= 0 and p.y < self.w

        return list(
            filter(
                good,
                [
                    Point(point.x, point.y - 1),
                    Point(point.x, point.y + 1),
                    Point(point.x - 1, point.y),
                    Point(point.x + 1, point.y),
                ],
            )
        )

    def count_nine(self, point: Point) -> int:
        assert self.data[point.x][point.y] == 0

        currents = {point}
        for score in range(1, 10):
            new_currents = set()
            for p in currents:
                for adj in self.adjacents(p):
                    if self.data[adj.x][adj.y] == score:
                        new_currents.add(adj)
            currents = new_currents

        return len(currents)

    def count_nine2(self, point: Point) -> int:
        assert self.data[point.x][point.y] == 0

        score_map: dict[Point, int] = defaultdict(lambda: 0)
        score_map[point] = 1

        currents = {point}
        for score in range(1, 10):
            new_currents = set()
            for p in currents:
                for adj in self.adjacents(p):
                    if self.data[adj.x][adj.y] == score:
                        score_map[adj] += score_map[p]
                        new_currents.add(adj)
            currents = new_currents

        s = 0
        for p in currents:
            s += score_map[p]
        return s


def read_map() -> Map:
    with utils.read(day=10, sample=False) as f:
        data = [
            list(map(int, line.strip())) for line in f.readlines() if len(line.strip())
        ]
    return Map(data)


def part1() -> None:
    m = read_map()

    scores = 0
    for zero in m.zeros():
        scores += m.count_nine(zero)

    print(scores)


def part2() -> None:
    m = read_map()

    scores = 0
    for zero in m.zeros():
        scores += m.count_nine2(zero)

    print(scores)


if __name__ == "__main__":
    part1()
    part2()
